### ISSUE TITLE
How to speed up server creation in vCD with Terraform.

### ISSUE THREAD

**Issue by cemkayar** [vmware/terraform-provider-vcd#579]
Hello @Vaidotas,

We opened a case to vmware to learn the parallel support of VCD. In response, the support engineer provided your name and address link. Could you share an example of terraform that shows how to create many vm at the same time, ie parallel run?

BR
Thanks

**Comment by vaidotas** [comment]
> Hello @vaidotas,
> 
> We opened a case to vmware to learn the parallel support of VCD. In response, the support engineer provided your name and address link. Could you share an example of terraform that shows how to create many vm at the same time, ie parallel run?
> 
> BR
> Thanks

Hey @cemkayar , 

You missed names, I shouldnt be here:)

**Comment by cemkayar** [comment]
Sorry @vaidotas 
@vbauzysvmware that would be the right person. :)

**Comment by vbauzysvmware** [comment]
Hi  @cemkayar,

We have idea how you can improve VM creation time. It's not ideal, but still a possibility to use. Idea is a way of creating VM in separate vApps and then merging them to right one. VMs in separate vApps are created in parallel. So we can have this steps:

1. Create VMs parallely:

```
resource "vcd_vapp" "web_temp1" {
  name = "web_temp1"
}
resource "vcd_vapp" "web_temp2" {
  name = "web_temp1"
}
resource "vcd_vapp" "web_temp3" {
  name = "web_temp1"
}
resource "vcd_vapp" "web_final" {
  name = "web_final"
}
resource "vcd_vapp_vm" "web1" {
  vapp_name     = vcd_vapp.web_temp1.name
...
}
resource "vcd_vapp_vm" "web2" {
  vapp_name     = vcd_vapp.web_temp2.name
...
}
resource "vcd_vapp_vm" "web3" {
  vapp_name     = vcd_vapp.web_temp3.name
...
}
```

2. Merge all 3 VM to `web_final` by using API call or UI `move to` functionality. This is very cheap actions and happens fast. As result all VMs will be created in parallel and moved to 4th vApp.

3. Update terraform configuration file to reference right vApp and delete temporary ones:
```
resource "vcd_vapp" "web_final" {
  name = "web_final"
}
resource "vcd_vapp_vm" "web1" {
  vapp_name     = vcd_vapp.web_final.name
...
}
resource "vcd_vapp_vm" "web2" {
  vapp_name     = vcd_vapp.web_final.name
...
}
resource "vcd_vapp_vm" "web3" {
  vapp_name     = vcd_vapp.web_final.name
...
}
```

**Comment by vbauzysvmware** [comment]
For Ref of #507 and #530

**Comment by cemkayar** [comment]
Hi @vbauzysvmware,

Thank you for code sample. As you wrote, this is not ideal, especially if we want to create 50 vm at the same time.
Is there any improvement on this in the roadmap?

**Comment by vbauzysvmware** [comment]
Hi @cemkayar only possible improvement is using scripting likes this, we have option just to look for possible improvements of current provider capabilities to make it less manual as I provided in example, if it would help.

**Comment by jpbuecken** [comment]
Because it is not mentioned in the original question: Is it important that all VMs end up in the same vapp?

If the application does not require it, we use 1 vm per vapp, because some changes to the vapp require that the vapp is powered off (and in turn, all vms in the vapp needs to be powered off)

If vapp is not required, maybe https://github.com/vmware/terraform-provider-vcd/issues/399 will help (create standalone vms)

Or the count feature of terraform. Did not test the following, but it should be similar to
```
resource "vcd_vapp" "webs" {
  count=50
  name = string("vappweb",count.index)
}
resource "vcd_vapp_vm" "web" {
  count=50
  vapp_name     = vcd_vapp.webs[count.index].name
...
}
```

But I'm not sure if terraform starts with vcd_vapp_vm.web[0] if the array vcd_vapp.webs is not completed yet...
Otherwise all vapps are created and then it begins with the vms. That is maybe not the fasted way, anyway the deployment of a vapp is not that time consuming, deployment of the more time consuming VMs is parallized

**Comment by vbauzysvmware** [comment]
@jpbuecken thank you for sharing your insights.

**Comment by jpbuecken** [comment]
@fatihbm: Can you share some thoughts why my comment https://github.com/vmware/terraform-provider-vcd/issues/579?_pjax=%23js-repo-pjax-container#issuecomment-717369934 is not useful? I'm curious.

- Original poster did not mention an answer to "Is it important that all VMs end up in the same vapp?"
- Question of OP did not give any clue if he is aware of capabilities of terraform

So whats wrong with an answer that mentions further enhancements (shorter hcl), other ideas (feature standalone vms) and some capabilities of terraform (count, foreach) that may be useful for OP. The true limitation is that VCD only allows one operation in one vApp per time. So everything here will be a workaround.


Anyway, if "Is it important that all VMs end up in the same vapp?" is mandatory, the step
 > Merge all 3 VM to web_final by using API call or UI move to functionality. 
from https://github.com/vmware/terraform-provider-vcd/issues/579?_pjax=%23js-repo-pjax-container#issuecomment-714231196
can maybe done with terraform as well: Use an api call in a null-resource.

**Comment by fatihbm** [comment]
> @fatihbm: Can you share some thoughts why my comment #579 is not useful? I'm curious.
> 
> * Original poster did not mention an answer to "Is it important that all VMs end up in the same vapp?"
> * Question of OP did not give any clue if he is aware of capabilities of terraform
> 
> So whats wrong with an answer that mentions further enhancements (shorter hcl), other ideas (feature standalone vms) and some capabilities of terraform (count, foreach) that may be useful for OP. The true limitation is that VCD only allows one operation in one vApp per time. So everything here will be a workaround.
> 
> Anyway, if "Is it important that all VMs end up in the same vapp?" is mandatory, the step
> 
> > Merge all 3 VM to web_final by using API call or UI move to functionality.
> 
> from #579
> can maybe done with terraform as well: Use an api call in a null-resource.
Actually, the question and expectation are clear.  Anyway, I can answer like this.
Can you sure this limitation come from vcd? I don't sure. Because vcd allows parallel vm creation same in vapp from ui! Also vmware guys said 'we don't limit it inside vcd'. I think so this limitation just correct for terraform.
On the other hand, your suggestion may be used for need small daily operation capacity in most cases. But unfortunately, it is not useful for an environment where hundreds of vm are created or deleted every day. By the way, I wish there was no obligation to use vapp when creating vm with Terraform. Then none of these would be a problem.
I hope this comment more revealing thank you for spending time

**Comment by cemkayar** [comment]
Vapp is not important to us for now, we are going on with a vapp for each vm. In this way, we became rapidly scalable. We don't need to merge vms under a vapp again.

**Comment by jpbuecken** [comment]
> Can you sure this limitation come from vcd? I don't sure. Because vcd allows parallel vm creation same in vapp from ui! Also vmware guys said 'we don't limit it inside vcd'. I think so this limitation just correct for terraform.

I based by answer from earlier answer in https://github.com/vmware/terraform-provider-vcd/issues/507 (already referenced here). I'm not aware if newer VCD dropped this limitation.

> By the way, I wish there was no obligation to use vapp when creating vm with Terraform. Then none of these would be a problem.

So you don't need a vApp. In that case the feature to create vms without vapp will help you? 
For this case, I referenced this issue in my comment you disliked, feel free to add your "like" to that issue:
> If (direct access to) the vapp is not required, maybe #399 will help (create standalone vms)

**Comment by lvirbalas** [comment]
Hello,

> I based by answer from earlier answer in #507 (already referenced here). I'm not aware if newer VCD dropped this limitation.

It hasn't.

> Vapp is not important to us for now, we are going on with a vapp for each vm. In this way, we became rapidly scalable. We don't need to merge vms under a vapp again.

We are hoping to add standalone VM feature to 3.2.0:
https://github.com/vmware/terraform-provider-vcd/issues/399#issuecomment-745197577

> Can you sure this limitation come from vcd? I don't sure. Because vcd allows parallel vm creation same in vapp from ui! 

Please, could you explain, share some screenshot how you do that in UI? This may actually help.

**Comment by dataclouder** [comment]
Released in [3.2.0](https://registry.terraform.io/providers/vmware/vcd/latest)
See a full example of usage [here](https://gist.github.com/dataclouder/20bea04eeb59c4dc821fc40edf426c0c)

Documentation:
*  [resource](https://registry.terraform.io/providers/vmware/vcd/latest/docs/resources/vm) 
* [data source](https://registry.terraform.io/providers/vmware/vcd/latest/docs/data-sources/vm).

A quick benchmark:
```
1 vApp with 10 VMs

          with network     bare VM
          ------------     -----------
Apply     21m  5.097s      12m 24.084s
Destroy    5m 47.492s       4m 26.281s

10 standalone VM

          with network     bare VM
          ------------     -----------
Apply      2m 58.690s       1m 31.818s
Destroy    1m 15.381s       0m 40.361s
```

Note: times may be a lot better or a lot worse depending on the hardware + network used.

You can quickly try the multiple VM creation with the trick below.

```hcl
resource "vcd_vm" "TestMultiStandalone" {
  for_each      = toset([ "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten"])
  org           = "datacloud"
  vdc           = "vdc-datacloud"
  name          = each.value
  catalog_name  = "cat-datacloud"
  template_name = "photon-hw11"
  memory        = 1024
  cpus          = 1
  power_on      = false
}
```
and the corresponding vApp+VM
```hcl

resource "vcd_vapp" "multiVmVapp" {
  org  = "datacloud"
  vdc  = "vdc-datacloud"
  name = "multiVmVapp"
  power_on = false
}
resource "vcd_vapp_vm" "testMultiVappVm" {
  for_each      = toset([ "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten" ])
  org           = "datacloud"
  vdc           = "vdc-datacloud"
  vapp_name     = "multiVmVapp"
  name          = each.value
  catalog_name  = "cat-datacloud"
  template_name = "photon-hw11"
  memory        = 1024
  cpus          = 1
  power_on      = false
}
```
